### PR TITLE
Feature: Highlighted Code Copy Blocks

### PR DESCRIPTION
**Pull request by renebrandel**
#### Description of changes:

Adds support for highlighted code blocks to be individually copied. In customer user study, every customer had trouble copying exactly the highlighted area. Doing a pass on highlighted code blocks revealed that the almost every single one required customers to explicitly copy the selected area to be successful as other areas of the code are truncated. 

Behavior of the highlighted code blocks:
- only copies the content within the highlighted area
- user can click anywhere on the highlighted area to copy
- user can use keyboard navigation to focus on a highlighted code block
- if code block has highlighted areas, then the global copy button is disabled

https://github.com/aws-amplify/docs/assets/4989523/1e7d039b-02f8-423b-bd56-6795dac37aa9

### Instructions

**If this PR should not be merged upon approval for any reason, please submit as a DRAFT**

Which product(s) are affected by this PR (if applicable)?
- [ ] amplify-cli
- [ ] amplify-ui
- [ ] amplify-studio
- [ ] amplify-hosting
- [ ] amplify-libraries

Which platform(s) are affected by this PR (if applicable)?
- [ ] JS
- [ ] Swift
- [ ] Android
- [ ] Flutter
- [ ] React Native

**Please add the product(s)/platform(s) affected to the PR title**

#### Checks

- [ ] Does this PR conform to [the styleguide](https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md)?

- [ ] Does this PR include filetypes other than markdown or images? Please add or update unit tests accordingly.

- [ ] Are any files being deleted with this PR? If so, have the needed redirects been created?

- [ ] Are all links in MDX files using the MDX link syntax rather than HTML link syntax? <br /> 
      _ref: MDX: `[link](https://docs.amplify.aws/)` 
            HTML: `<a href="https://docs.amplify.aws/">link</a>`_

### When this PR is ready to merge, please check the box below
- [ ] Ready to merge

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: src/components/MDXComponents/MDXCode.tsx [hunk 1]
```diff
@@ -30,6 +30,18 @@ const addVersions = (code: string) => {
   return code;
 };
 
+const hasHighlights = (code: string): boolean => {
+  const highlightableStrings = [
+    '// highlight-start',
+    '// highlight-end',
+    '// highlight-line-start'
```
Comment (timngyn) [inline src/components/MDXComponents/MDXCode.tsx:37]: Do we need to also add `// highlight-next-line` ?
```suggestion
    '// highlight-start',
    '// highlight-end',
    '// highlight-line-start',
    '// highlight-next-line'
```

File: src/styles/code.scss [hunk 2]
```diff
@@ -131,6 +131,40 @@ code:not([class]) {
...
+  display: block;
+}
+
+.highlight-copy-block:focus .line-highlight::before {
+  background-color: var(--amplify-colors-primary-80);
+}
+
+.highlight-copy-block:hover .line-highlight::before {
+  background-color: var(--amplify-colors-primary-90);
+}
```
[the remark this answers is not in the document]
Comment (timngyn) [reply src/styles/code.scss:166]: Maybe `--amplify-colors-primary-20` for focus and `--amplify-colors-primary-10` for hover?

**Comment by renebrandel** [comment]
https://github.com/aws-amplify/docs/assets/4989523/6ec005e6-ba48-4874-9eb5-4b072568fcfc

@timngyn  added dark mode colors

**Comment by renebrandel** [comment]
@timngyn  fixed tests

**Review by timngyn — APPROVED** [review]

**Review by swaminator — APPROVED** [review]

**Pull request merged**